### ISSUE TITLE
Could not get the id of the latest version for subject - Pyspark 2.4.4 and Abris 3.2.1

### ISSUE THREAD

**Issue by vphutane** [AbsaOSS/ABRiS#140]
I want to use to_avro and publish my schema to schema registry if not exist. It gives me error saying " za.co.absa.abris.avro.read.confluent.SchemaManagerException: Could not get the id of the latest version for subject 'canonicalaccount-dataingestion.resources.accountcreation'". Below is the to_avro method. I am using confluent kafka, pyspark 2.4.4. In the schema registry config, I tried providing everything for schemaRegistryConfig. Does it require anyother parameter.

df \
            .withColumn("value", struct([df[x] for x in df.columns])) \
            .select("value", to_avro("value", output_topic_name, SCHEMA_REGISTRY_URL)) \
            .writeStream \
            .format("kafka") \

def to_avro(col, topic, schema_registry_url):
        jvm_gateway = SparkContext._active_spark_context._gateway.jvm
        abris_avro = jvm_gateway.za.co.absa.abris.avro
        naming_strategy = getattr(
            getattr(abris_avro.read.confluent.SchemaManager, "SchemaStorageNamingStrategies$"),
            "MODULE$"
        ).TOPIC_RECORD_NAME()

        schema_registry_config_dict = {
            "schema.registry.url": schema_registry_url,
            "schema.registry.topic": topic,
            "{col}.schema.id".format(col=col): "latest",
            "{col}.schema.naming.strategy".format(col=col): naming_strategy,
            "{col}.schema.record.name".format(col=col): "accountcreation",
            "{col}.schema.record.namespace".format(col=col): "dataingestion.resources",
            "{col}.schema.name".format(col=col): "accountcreation",
            "{col}.schema.namespace".format(col=col): "dataingestion.resources"
        }

        conf_map = getattr(getattr(jvm_gateway.scala.collection.immutable.Map, "EmptyMap$"), "MODULE$")
        for k, v in schema_registry_config_dict.items():
            conf_map = getattr(conf_map, "$plus")(jvm_gateway.scala.Tuple2(k, v))

        return Column(abris_avro.functions.to_confluent_avro(_to_java_column(col), conf_map))

**Comment by cerveada** [comment]
`>> I want to use to_avro and publish my schema to schema registry if not exist.`

Hello, currently there is no **"if not exist"** functionality in Abris. You can either use already existing schema in the registry or generate one or provide one yourself.

If you provide schema id Abris will try to find the schema in registry and if it's not there it's an error.

Check the documentation for more info: https://github.com/AbsaOSS/ABRiS/blob/master/documentation/confluent-avro-documentation.md#spark-to-avro

**Comment by vphutane** [comment]
Thanks....actually what I want to ask is if the schema does not exist in the schema registry, then abris should generate schema from the dataframe structure as per the documentation. However at moment it is not automatically generating the schema from the dataframe structure and registering. Don't know if I missed any configuration entry in schema_registry_config_dict or I missed any def / method to call before invoking to_avro

**Comment by cerveada** [comment]
Sorry for late response, there is no such feature, but as a workaround you can do this:

You have to found yourself whether the schema exists and then call Abris with different configurations:
- if schema exist you provide the coordinates and id and abris will use it
- if schema doesn't exist you must not provide id or version in config - if they are not there abris takes it as a signal it should generate the schema


To find out if schema exists is not Abris feature currently (we're planing to add it in future), but its private method in `SchemaManager` called `exists` so you can take a look and copy the code:
https://github.com/AbsaOSS/ABRiS/blob/master/src/main/scala/za/co/absa/abris/avro/read/confluent/SchemaManager.scala

**Comment by vphutane** [comment]
Thank you very much and I truly appreciate for your help. Both the links were really helpful. I was successfully able to play around and get the schema automatically created from dataframe and register the schema with schema registry. I was also able to put multiple events i.e. different record types and register their schema for those events into the same Topic. For all those different events ABRIS registered their schema as v1....This is what I wanted to achieve. 

I am now doing another test with change in structure in dataframe i.e. I have introduced 1 additional column in my dataframe and want to check whether ABRIS can register this new schema as v2 for one of the event. However my job failed with error "java.security.InvalidParameterException: Schema could not be registered for subject 'canonicalaccount-mybank.account.closed_account'. Make sure that the Schema Registry is available, the parameters are correct and the schemas are compatible". Want to check with you is this feasible ? ....or do I need to manually update the schema in the registry.

**Comment by cerveada** [comment]
I think this should work. 
- Do you provide the updated schema or do you let Abris generate it? 
- Could you provide the code that is causing this?

**Comment by vphutane** [comment]
I didn't provide the schema. I was testing if ABRIS can generate it looking to the structure from dataframe. It did generate the schema from dataframe correctly for version v1, but later i applied a change in the dataframe structure by adding a column, I was expecting ABRIS to read the changed dataframe structure and create a new v2 structure and register...but it errored. Below is the write_avro method 

**invoking process**
         df \
            .withColumn("value", struct([df[x] for x in df.columns])) \
            .withColumn("key", lit(query_name)) \
            .select(write_avro("key", output_topic_name, SCHEMA_REGISTRY_URL, query_name).alias("key"),
                    write_avro("value", output_topic_name, SCHEMA_REGISTRY_URL, query_name).alias(
                        "value")) 
 
**### write-avro method**
    def write_avro(col, topic, schema_registry_url, query_name):
        jvm_gateway = SparkContext._active_spark_context._gateway.jvm
        abris_avro = jvm_gateway.za.co.absa.abris.avro

        key_naming_strategy = getattr(
            getattr(abris_avro.read.confluent.SchemaManager, "SchemaStorageNamingStrategies$"),
            "MODULE$"
        ).TOPIC_NAME()

        value_naming_strategy = getattr(
            getattr(abris_avro.read.confluent.SchemaManager, "SchemaStorageNamingStrategies$"),
            "MODULE$"
        ).TOPIC_RECORD_NAME()

        value_registry_config_dict = {
            "schema.registry.url": schema_registry_url,
            "schema.registry.topic": topic,
            "value.schema.name": query_name,
            "value.schema.namespace": "ben.account"
        }

        key_registry_config_dict = {
            "schema.registry.url": schema_registry_url,
            "schema.registry.topic": topic,
            "key.schema.name": "account",
            "key.schema.namespace": "ben.account"
        }

        key_registry_config_name_dict = {
            "key.schema.naming.strategy": key_naming_strategy
        }

        value_registry_config_name_dict = {
            "value.schema.naming.strategy": value_naming_strategy
        }

        schema_registry_config_dict_key = {**key_registry_config_dict, **key_registry_config_name_dict}
        schema_registry_config_dict_value = {**value_registry_config_dict, **value_registry_config_name_dict}

        conf_map = getattr(getattr(jvm_gateway.scala.collection.immutable.Map, "EmptyMap$"), "MODULE$")

        if col == "key":
            for k, v in schema_registry_config_dict_key.items():
                conf_map = getattr(conf_map, "$plus")(jvm_gateway.scala.Tuple2(k, v))
        elif col == "value":
            for k, v in schema_registry_config_dict_value.items():
                conf_map = getattr(conf_map, "$plus")(jvm_gateway.scala.Tuple2(k, v))

        return Column(abris_avro.functions.to_confluent_avro(_to_java_column(col), conf_map))

**Comment by sivasai-quartic** [comment]
Hi @cerveada 

I'm facing the issue while fetching the schema from the schema registry. In my case schema is available in schema registry. 
exception is caused because of this  org.apache.hadoop.fs.FsUrlConnection cannot be cast to java.net.HttpURLConnection

```
Exception in thread "main" za.co.absa.abris.avro.read.confluent.SchemaManagerException: Could not get the id of the latest version for subject 'schema-name-key'
	at za.co.absa.abris.avro.read.confluent.SchemaManager.getLatestVersionId(SchemaManager.scala:59)
	at za.co.absa.abris.avro.read.confluent.SchemaManager.resolveSchemaId(SchemaManager.scala:40)
	at za.co.absa.abris.avro.read.confluent.SchemaManager.schemaId$lzycompute(SchemaManager.scala:34)
	at za.co.absa.abris.avro.read.confluent.SchemaManager.schemaId(SchemaManager.scala:34)
	at za.co.absa.abris.avro.read.confluent.SchemaManager.downloadSchema(SchemaManager.scala:76)
	at za.co.absa.abris.avro.functions$.from_confluent_avro(functions.scala:87)
	at 
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:928)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1007)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1016)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.ClassCastException: org.apache.hadoop.fs.FsUrlConnection cannot be cast to java.net.HttpURLConnection
	at io.confluent.kafka.schemaregistry.client.rest.RestService.sendHttpRequest(RestService.java:176)
	at io.confluent.kafka.schemaregistry.client.rest.RestService.httpRequest(RestService.java:256)
	at io.confluent.kafka.schemaregistry.client.rest.RestService.getLatestVersion(RestService.java:515)
	at io.confluent.kafka.schemaregistry.client.rest.RestService.getLatestVersion(RestService.java:507)
	at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getLatestSchemaMetadata(CachedSchemaRegistryClient.java:275)
	at za.co.absa.abris.avro.read.confluent.SchemaManager.$anonfun$getLatestVersionId$2(SchemaManager.scala:56)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.java:23)
	at scala.util.Try$.apply(Try.scala:213)
	at za.co.absa.abris.avro.read.confluent.SchemaManager.getLatestVersionId(SchemaManager.scala:56)
	... 23 more

```

**Comment by cerveada** [comment]
@sivasai-quartic Hello, each issue should have its own ticket. Unless you a have comment about the same issue or your problem is closely related create new ticket.

**Comment by cerveada** [comment]
@vphutane The problem is schema compatibility. The new field must be declared optional or default value must be provided. There is also more types: backwards, forwards and full. If I am not mistaken this is setting in the registry. 

Abris doesn't know how do you want to achieve that compatibility. For example if there is default value how is Abris supposed to know what should be the default?

So it will generate the schema for you, but without any regards for compatibility.

**Comment by vphutane** [comment]
Thanks @cerveada ....As suggested above I have now changed my code to read schema from .avsc file. However after changing to  read schema from .avsc file, it gives error saying "org.apache.spark.sql.avro.IncompatibleSchemaException: Cannot convert Catalyst type StringType to Avro type". I am just reading the avsc file and converting the contents into string. The avsc file and output from spark dataframe is as below

avsc file
{
  "type": "record",
  "name": "account_creation",
  "namespace": "ben.account",
  "fields": [
    {
      "name": "event_name",
      "type": "string"
    },
    {
      "name": "acct_id_id",
      "type": "string"
    },
    {
      "name": "acct_dsply_nme",
      "type": "string"
    },
    {
      "name": "acct_nick_nme",
      "type": "string"
    },
    {
      "name": "acct_opn_stat",
      "type": "string"
    },
    {
      "name": "acct_prod_cat",
      "type": "string"
    },
    {
      "name": "acct_brnd_nme",
      "type": "string"
    }
  ]
}

Also below is my spark dataframe
Read Streaming Data
root
 |-- event_name: string (nullable = false)
 |-- acct_id_id: string (nullable = false)
 |-- acct_dsply_nme: string (nullable = false)
 |-- acct_nick_nme: string (nullable = false)
 |-- acct_opn_stat: string (nullable = false)
 |-- acct_prod_cat: string (nullable = false)
 |-- acct_brnd_nme: string (nullable = false)

the schema structure in avsc file and in my spark dataframe, they both are in sync and compatible. However I am still receiving the catalyst error saying 

20/08/14 18:08:59 ERROR MicroBatchExecution: Query account_creation [id = 32e3284c-7c64-4554-87a1-bc906fb9c33d, runId = 1e404e6c-c58a-4292-8a76-61e75c2db75e] terminated with error
org.apache.spark.sql.avro.IncompatibleSchemaException: Cannot convert Catalyst type StringType to Avro type {"type":"record","name":"account_creation","namespace":"ben.account","fields":[{"name":"event_name","type":"string"},{"name":"acct_id_id","type":"string"},{"name":"acct_dsply_nme","type":"string"},{"name":"acct_nick_nme","type":"string"},{"name":"acct_opn_stat","type":"string"},{"name":"acct_prod_cat","type":"string"},{"name":"acct_brnd_nme","type":"string"}]}.
        at org.apache.spark.sql.avro.AvroSerializer.org$apache$spark$sql$avro$AvroSerializer$$newConverter(AvroSerializer.scala:197)
        at org.apache.spark.sql.avro.AvroSerializer.<init>(AvroSerializer.scala:54)
        at za.co.absa.abris.avro.sql.CatalystDataToAvro.serializer$lzycompute(CatalystDataToAvro.scala:44)
        at za.co.absa.abris.avro.sql.CatalystDataToAvro.serializer(CatalystDataToAvro.scala:43)
        at za.co.absa.abris.avro.sql.CatalystDataToAvro.nullSafeEval(CatalystDataToAvro.scala:49)
        at org.apache.spark.sql.catalyst.expressions.UnaryExpression.eval(Expression.scala:393)
        at org.apache.spark.sql.catalyst.optimizer.ConstantFolding$$anonfun$apply$1$$anonfun$applyOrElse$1.applyOrElse(expressions.scala:52)
        at org.apache.spark.sql.catalyst.optimizer.ConstantFolding$$anonfun$apply$1$$anonfun$applyOrElse$1.applyOrElse(expressions.scala:45)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:258)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:258)
        at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:69)
        at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:257)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformDown$1.apply(TreeNode.scala:263)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformDown$1.apply(TreeNode.scala:263)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$4.apply(TreeNode.scala:328)
        at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:186)
        at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:326)
        at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:263)
        at org.apache.spark.sql.catalyst.plans.QueryPlan$$anonfun$transformExpressionsDown$1.apply(QueryPlan.scala:83)
        at org.apache.spark.sql.catalyst.plans.QueryPlan$$anonfun$transformExpressionsDown$1.apply(QueryPlan.scala:83)
        at org.apache.spark.sql.catalyst.plans.QueryPlan$$anonfun$1.apply(QueryPlan.scala:105)
        at org.apache.spark.sql.catalyst.plans.QueryPlan$$anonfun$1.apply(QueryPlan.scala:105)
        at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:69)
        at org.apache.spark.sql.catalyst.plans.QueryPlan.transformExpression$1(QueryPlan.scala:104)
        at org.apache.spark.sql.catalyst.plans.QueryPlan.org$apache$spark$sql$catalyst$plans$QueryPlan$$recursiveTransform$1(QueryPlan.scala:116)
        at org.apache.spark.sql.catalyst.plans.QueryPlan$$anonfun$org$apache$spark$sql$catalyst$plans$QueryPlan$$recursiveTransform$1$2.apply(QueryPlan.scala:121)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at org.apache.spark.sql.catalyst.plans.QueryPlan.org$apache$spark$sql$catalyst$plans$QueryPlan$$recursiveTransform$1(QueryPlan.scala:121)
        at org.apache.spark.sql.catalyst.plans.QueryPlan$$anonfun$2.apply(QueryPlan.scala:126)
        at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:186)
        at org.apache.spark.sql.catalyst.plans.QueryPlan.mapExpressions(QueryPlan.scala:126)
        at org.apache.spark.sql.catalyst.plans.QueryPlan.transformExpressionsDown(QueryPlan.scala:83)
        at org.apache.spark.sql.catalyst.optimizer.ConstantFolding$$anonfun$apply$1.applyOrElse(expressions.scala:45)
        at org.apache.spark.sql.catalyst.optimizer.ConstantFolding$$anonfun$apply$1.applyOrElse(expressions.scala:44)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:258)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:258)
        at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:69)
        at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:257)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDown(LogicalPlan.scala:29)
        at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$class.transformDown(AnalysisHelper.scala:149)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDown(LogicalPlan.scala:29)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDown(LogicalPlan.scala:29)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformDown$1.apply(TreeNode.scala:263)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformDown$1.apply(TreeNode.scala:263)
        at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$4.apply(TreeNode.scala:328)
        at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:186)
        at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:326)
        at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:263)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDown(LogicalPlan.scala:29)
        at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$class.transformDown(AnalysisHelper.scala:149)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDown(LogicalPlan.scala:29)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDown(LogicalPlan.scala:29)
        at org.apache.spark.sql.catalyst.trees.TreeNode.transform(TreeNode.scala:247)
        at org.apache.spark.sql.catalyst.optimizer.ConstantFolding$.apply(expressions.scala:44)
        at org.apache.spark.sql.catalyst.optimizer.ConstantFolding$.apply(expressions.scala:43)
        at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1$$anonfun$apply$1.apply(RuleExecutor.scala:87)
        at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1$$anonfun$apply$1.apply(RuleExecutor.scala:84)
        at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
        at scala.collection.immutable.List.foldLeft(List.scala:84)
        at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1.apply(RuleExecutor.scala:84)
        at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1.apply(RuleExecutor.scala:76)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:76)
        at org.apache.spark.sql.execution.streaming.IncrementalExecution.optimizedPlan$lzycompute(IncrementalExecution.scala:77)
        at org.apache.spark.sql.execution.streaming.IncrementalExecution.optimizedPlan(IncrementalExecution.scala:77)
        at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:73)
        at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:69)
        at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:78)
        at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:78)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$org$apache$spark$sql$execution$streaming$MicroBatchExecution$$runBatch$4.apply(MicroBatchExecution.scala:528)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$org$apache$spark$sql$execution$streaming$MicroBatchExecution$$runBatch$4.apply(MicroBatchExecution.scala:519)
        at org.apache.spark.sql.execution.streaming.ProgressReporter$class.reportTimeTaken(ProgressReporter.scala:351)
        at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:58)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.org$apache$spark$sql$execution$streaming$MicroBatchExecution$$runBatch(MicroBatchExecution.scala:519)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1$$anonfun$apply$mcZ$sp$1.apply$mcV$sp(MicroBatchExecution.scala:198)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1$$anonfun$apply$mcZ$sp$1.apply(MicroBatchExecution.scala:166)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1$$anonfun$apply$mcZ$sp$1.apply(MicroBatchExecution.scala:166)
        at org.apache.spark.sql.execution.streaming.ProgressReporter$class.reportTimeTaken(ProgressReporter.scala:351)
        at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:58)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1.apply$mcZ$sp(MicroBatchExecution.scala:166)
        at org.apache.spark.sql.execution.streaming.ProcessingTimeExecutor.execute(TriggerExecutor.scala:56)
        at org.apache.spark.sql.execution.streaming.MicroBatchExecution.runActivatedStream(MicroBatchExecution.scala:160)
        at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:281)
        at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:193)
2020-08-14 18:09:00,028 — dataingestion.jobs.ingester — ERROR — main:56 — Exception while waiting for query to end

Below is how I am reading the avsc file in my pyspark code.

schema_path = "/home/xxx/development/ingestion/account_creation.avsc"
        with open(schema_path) as schema_file:
            schema_string = schema_file.read()
        schema_file.close()

Column(abris_avro.functions.to_confluent_avro(_to_java_column(col), schema_string, conf_map))

Earlier I tried with 1 column but still same issue. I now included 7 columns. I am really stuck at this point. Can you please assist with your expertise. Truly appreciate for your help.

**Comment by cerveada** [comment]
It seems to me you are providing just one column here: 
```
Column(abris_avro.functions.to_confluent_avro(_to_java_column(col), schema_string, conf_map))
```

The schema contains several columns, so the data must match.

In Scala, I would put all columns to struct like that:
```
  val allColumns = struct(dataFrame.columns.head, dataFrame.columns.tail: _*)
  dataFrame.select(to_confluent_avro(allColumns, schemaString, config) as 'value)
```

**Comment by vphutane** [comment]
Hi @cerveada , _to_java_column(**### col**) is a struct of several columns. In Pyspark it is done as struct([df[x] for x in df.columns]). I have demonstrated below. This is similar to the one in scala allColumns  you provided above.

>>> df = spark.read.format("csv").option("header", "true").option("inferSchema", "true").option("model", "DROPMALFORMED").load("file:///home/vinodp/development/ingestion/sampledata.csv")
>>> df.printSchema()
root
 |-- id: integer (nullable = true)
 |-- name: string (nullable = true)
 |-- email: string (nullable = true)

>>> all_columns = struct([df[x] for x in df.columns])
>>> print(all_columns)
Column<b'named_struct(id, id, name, name, email, email)'>
>>>

**Comment by ottensjors** [comment]
> Thank you very much and I truly appreciate for your help. Both the links were really helpful. I was successfully able to play around and get the schema automatically created from dataframe and register the schema with schema registry. I was also able to put multiple events i.e. different record types and register their schema for those events into the same Topic. For all those different events ABRIS registered their schema as v1....This is what I wanted to achieve.
> 
> I am now doing another test with change in structure in dataframe i.e. I have introduced 1 additional column in my dataframe and want to check whether ABRIS can register this new schema as v2 for one of the event. However my job failed with error "java.security.InvalidParameterException: Schema could not be registered for subject 'canonicalaccount-mybank.account.closed_account'. Make sure that the Schema Registry is available, the parameters are correct and the schemas are compatible". Want to check with you is this feasible ? ....or do I need to manually update the schema in the registry.

hi @vphutane  - im currenty working on automatically creating a schema from a dataframe and registring it to schema-registry. however, the schema subject does not appear in the confluent schema-registry.  would you be so kind to share your code for config-map etc? I think i'm missing something small.

regards Sjors Otten

**Comment by vphutane** [comment]
@ottensjors

The below is a pyspark code. But remember if you'r structures are changing over a period of time then below process will not help. You will need to provide avro schema file in the to_confluent_avro

invoking process
df
.withColumn("value", struct([df[x] for x in df.columns]))
.withColumn("key", lit(query_name))
.select(write_avro("key", output_topic_name, SCHEMA_REGISTRY_URL, query_name).alias("key"),
write_avro("value", output_topic_name, SCHEMA_REGISTRY_URL, query_name).alias(
"value"))

### write-avro method
def write_avro(col, topic, schema_registry_url, query_name):
jvm_gateway = SparkContext._active_spark_context._gateway.jvm
abris_avro = jvm_gateway.za.co.absa.abris.avro

    key_naming_strategy = getattr(
        getattr(abris_avro.read.confluent.SchemaManager, "SchemaStorageNamingStrategies$"),
        "MODULE$"
    ).TOPIC_NAME()

    value_naming_strategy = getattr(
        getattr(abris_avro.read.confluent.SchemaManager, "SchemaStorageNamingStrategies$"),
        "MODULE$"
    ).TOPIC_RECORD_NAME()

    value_registry_config_dict = {
        "schema.registry.url": schema_registry_url,
        "schema.registry.topic": topic,
        "value.schema.name": query_name,
        "value.schema.namespace": "ben.account"
    }

    key_registry_config_dict = {
        "schema.registry.url": schema_registry_url,
        "schema.registry.topic": topic,
        "key.schema.name": "account",
        "key.schema.namespace": "ben.account"
    }

    key_registry_config_name_dict = {
        "key.schema.naming.strategy": key_naming_strategy
    }

    value_registry_config_name_dict = {
        "value.schema.naming.strategy": value_naming_strategy
    }

    schema_registry_config_dict_key = {**key_registry_config_dict, **key_registry_config_name_dict}
    schema_registry_config_dict_value = {**value_registry_config_dict, **value_registry_config_name_dict}

    conf_map = getattr(getattr(jvm_gateway.scala.collection.immutable.Map, "EmptyMap$"), "MODULE$")

    if col == "key":
        for k, v in schema_registry_config_dict_key.items():
            conf_map = getattr(conf_map, "$plus")(jvm_gateway.scala.Tuple2(k, v))
    elif col == "value":
        for k, v in schema_registry_config_dict_value.items():
            conf_map = getattr(conf_map, "$plus")(jvm_gateway.scala.Tuple2(k, v))

    return Column(abris_avro.functions.to_confluent_avro(_to_java_column(col), conf_map))

**Comment by vphutane** [comment]
Hi @cerveada the above problem with to_confluent_avro is solved now. I was able to put multiple events of different record type in AVRO format using Schema Registry in a SINGLE topic using schema file for those different record types. The avsc schema file was correct...and there was no issue with catalyst sql. The issue was with SchemaStorageNamingStrategies name. I changed the same only for value part as TOPIC_RECORD_NAME and applied schema file.  For key i left it default without applying schema file....and it worked like charm.

I am now testing the from_confluent_avro to read multiple events of different record types from the same ONE topic......before I proceed want to check with you is this possible. Can I read multiple events of different record type from schema registry from a Single topic using ABRiS library ?.
